### PR TITLE
Fix submission monitoring

### PR DESCRIPTION
**Pull request by dvoet**
https://broadinstitute.atlassian.net/browse/GAWB-507

This PR refactors the submission monitoring to achieve several objectives:
1) remove concurrency which was leading to deadlocks
2) remove database queries that are not needed to monitor submissions
3) move the code toward what it will look like once cromwell has a batch end point

To address 1 and 3, we no longer use separate actors to monitor workflows. A single submission actor monitors all the workflows for that submission. We still have to monitor workflows with separate cromwell calls but those don't involve the rawls database.

To address 2, we don't load workflow or submission objects because we don't care about most of that information. Instead we query for SubmissionRecord and WorkflowRecord directly and do bulk queries as much as possible.
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala [hunk 1]
```diff
@@ -144,11 +147,11 @@ trait SubmissionComponent {
 
       uniqueResult[SubmissionRecord](findById(UUID.fromString(submission.submissionId))) flatMap {
         case None =>
-          val configId = uniqueResult(methodConfigurationQuery.findByName(
+          val configIdAction = uniqueResult[Long](methodConfigurationQuery.findByName(
             workspaceContext.workspaceId, submission.methodConfigurationNamespace, submission.methodConfigurationName).map(_.id))
 
           loadSubmissionEntityId(workspaceContext.workspaceId, submission.submissionEntity) flatMap { entityId =>
-            configId flatMap { configId =>
+            configIdAction flatMap { configId =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala:154]: this was just some unrelated confusion clean up

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/PerRequest.scala [hunk 2]
```diff
@@ -89,6 +89,9 @@ trait PerRequest extends Actor {
       case (statusCode: StatusCode, payload: Any) =>
         logResponse(statusCode, payload)
         r.withHttpResponseHeadersMapped(h => h ++ headers).complete(response)
+      case (statusCode: StatusCode) =>
+        logResponse(statusCode, None)
+        r.withHttpResponseHeadersMapped(h => h ++ headers).complete(response)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/PerRequest.scala:94]: this is an unrelated bug fix

File: src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala [hunk 3]
```diff
@@ -1,21 +1,28 @@
 package org.broadinstitute.dsde.rawls.jobexec
```
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala:1]: this is pretty much completely rewritten, the diff is not helpful

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+  private def execServiceStatus(workflowRec: WorkflowRecord)(implicit executionContext: ExecutionContext): Future[Option[WorkflowRecord]] = {
+    executionServiceDAO.status(workflowRec.externalId, getUserInfo).map(newStatus => {
+      if (newStatus.status != workflowRec.status) Option(workflowRec.copy(status = newStatus.status))
+      else None
+    })
+  }
+
+  private def execServiceOutputs(workflowRecOption: Option[WorkflowRecord])(implicit executionContext: ExecutionContext): Future[Option[(WorkflowRecord, Option[ExecutionServiceOutputs])]] = {
+    workflowRecOption match {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: Can this match be moved up to queryExecutionServiceForStatus?  It would be a little clearer if this method took a WorkflowRecord without the Option.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+      DBIO.successful(Unit)
+    } else {
+      for {
+        // load all the starting data
+        entitiesById <-         listWorkflowEntitiesById(workflowsWithOutputs, dataAccess)
+        outputExprepressions <- listMethodConfigOutputsForSubmission(dataAccess)
+        workspaceOption <-      getWorkspace(dataAccess)
+
+        // update the appropriate entities and workspace (in memory)
+        updatedEntitiesAndWorkspace = attachOutputs(workspaceOption.get, workflowsWithOutputs, entitiesById, outputExprepressions)
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: If the get is moved to getWorkspace (so it returns `ReadAction[Workspace]`) then any None.get errors will be closer to the immediate cause, so easier to debug if it ever happens.

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/TestDriverComponent.scala [hunk 6]
```diff
@@ -242,7 +242,7 @@ trait TestDriverComponent extends DriverComponent with DataAccess {
     val methodConfig2 = MethodConfiguration("dsde", "testConfig2", "Sample", Map("ready"-> AttributeString("true")), Map("param1"-> AttributeString("foo")), Map("out1" -> AttributeString("bar"), "out2" -> AttributeString("splat")), MethodRepoMethod(wsName.namespace, "method-a", 1))
     val methodConfig3 = MethodConfiguration("dsde", "testConfig", "Sample", Map("ready"-> AttributeString("true")), Map("param1"-> AttributeString("foo"), "param2"-> AttributeString("foo2")), Map("out" -> AttributeString("bar")), MethodRepoMethod("ns-config", "meth1", 1))
 
-    val methodConfigEntityUpdate = MethodConfiguration("ns", "testConfig1", "Sample", Map(), Map(), Map("o1" -> AttributeString("this.foo")), MethodRepoMethod("ns-config", "meth1", 1))
+    val methodConfigEntityUpdate = MethodConfiguration("ns", "testConfig11", "Sample", Map(), Map(), Map("o1" -> AttributeString("this.foo")), MethodRepoMethod("ns-config", "meth1", 1))
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/TestDriverComponent.scala:245]: Was this intentional? (probably doesn't matter either way)

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSupervisor.scala [hunk 7]
```diff
@@ -16,13 +17,12 @@ import scala.concurrent.duration._
...
 
-  case class SubmissionStarted(workspaceName: WorkspaceName, submissionId: String, credential: Credential)
+  case class SubmissionStarted(workspaceName: WorkspaceName, submissionId: UUID, credential: Credential)
 
   def props(executionServiceDAO: ExecutionServiceDAO,
             datasource: SlickDataSource,
-            workflowPollInterval: Duration = 1 minutes,
-            submissionPollInterval: Duration = 30 minutes): Props = {
-    Props(new SubmissionSupervisor(executionServiceDAO, datasource, workflowPollInterval, submissionPollInterval))
+            submissionPollInterval: FiniteDuration = 1 minutes): Props = {
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSupervisor.scala:24]: is it worth pulling the duration out into config, so we can tune it per-environment? (brainstorming here, not suggesting)

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+  def checkOverallStatus(dataAccess: DataAccess)(implicit executionContext: ExecutionContext): ReadWriteAction[Boolean] = {
+    dataAccess.workflowQuery.listWorkflowRecsForSubmissionAndStatuses(submissionId, WorkflowStatuses.runningStatuses:_*) flatMap { workflowRecs =>
+      if (workflowRecs.isEmpty) {
+        dataAccess.submissionQuery.findById(submissionId).map(_.status).result.head.map { status =>
+          SubmissionStatuses.withName(status) match {
+            case SubmissionStatuses.Aborting => SubmissionStatuses.Aborted
+            case SubmissionStatuses.Submitted => SubmissionStatuses.Done
+            case _ => throw new RawlsException(s"submission $submissionId in unexpected state $status, expected Aborting or Submitted")
+          }
+        } flatMap { newStatus => dataAccess.submissionQuery.updateStatus(submissionId, newStatus) } map(_ => true)
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:179]: potentially add a debug log statement here, noting the date/time that we marked the submission finished? (I'd love to see that in the db table, but that's way out of scope for this PR)

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+  }
+
+  def handleOutputs(workflowsWithOutputs: Seq[(WorkflowRecord, ExecutionServiceOutputs)], dataAccess: DataAccess)(implicit executionContext: ExecutionContext) = {
+    if (workflowsWithOutputs.isEmpty) {
+      DBIO.successful(Unit)
+    } else {
+      for {
+        // load all the starting data
+        entitiesById <-         listWorkflowEntitiesById(workflowsWithOutputs, dataAccess)
+        outputExprepressions <- listMethodConfigOutputsForSubmission(dataAccess)
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: typo on "outputExprepressions" ... should be "outputExpressions"

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+        workspaceOption <-      getWorkspace(dataAccess)
+
+        // update the appropriate entities and workspace (in memory)
+        updatedEntitiesAndWorkspace = attachOutputs(workspaceOption.get, workflowsWithOutputs, entitiesById, outputExprepressions)
+
+        // save everything to the db
+        _ <- saveWorkspace(dataAccess, updatedEntitiesAndWorkspace)
+        _ <- saveEntities(dataAccess, workspaceOption.get, updatedEntitiesAndWorkspace)
+        _ <- saveErrors(updatedEntitiesAndWorkspace.collect { case Right(errors) => errors }, dataAccess)
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: what happens on error in one of these three updates? Can/should this be transaction-ized?

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 5 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: I can do it in the first instance mentioned but not here because the for comprehension is unwrapping things so getWorkspace(dataAccess) is not actually of type Option (although I could map it...)

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    }
+  }
+
+  def updateEntityAndWorkspace(entity: Entity, workspace: Workspace, workflowOutputs: Map[String, Attribute]): (Option[Entity], Option[Workspace]) = {
+    //Partition outputs by whether their attributes are entity attributes (begin with "this.") or workspace ones (implicitly; begin with "workspace.")
+    //This assumption (that it's either "this." or "workspace.") will be guaranteed by checking of the method config when it's imported; see DSDEEPB-1603.
+    //Yes I know this is a var but it's more terse this way.
+    var (entityAttributes, workspaceAttributes) = workflowOutputs.partition({ case (k, v) => k.startsWith("this.") })
+    entityAttributes = entityAttributes.map({ case (k, v) => (k.stripPrefix("this."), v) })
+    workspaceAttributes = workspaceAttributes.map({ case (k, v) => (k.stripPrefix("workspace."), v) })
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: I hate vars :)

```
val (partitionEntity, partitionWorkspace) = workflowOutputs.partition({ case (k, v) => k.startsWith("this.") })
val entityAttributes = partitionEntity.map({ case (k, v) => (k.stripPrefix("this."), v) })
val workspaceAttributes = partitionWorkspace.map({ case (k, v) => (k.stripPrefix("workspace."), v) })
```

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/TestDriverComponent.scala (hunk 6 above)
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/TestDriverComponent.scala:245]: it was intentional, the name overlaps with another method config in the test fixture

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSupervisor.scala (hunk 7 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSupervisor.scala:24]: it is written to be configurable but has not been a priority to push it all the way into config

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 8 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:179]: there is a debug log in the receive of the actor... do you think that is good enough?

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 10 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: it is in a transaction, the transaction demarcation is in handleStatusResponses

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 11 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: this is actually copied from the old code, it works so I left it alone but if you want me to change it I will

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+  import datasource.dataAccess.driver.api._
+
+  /**
+   * This function starts a monitoring pass
+   * @param executionContext
+   * @return
+   */
+  def queryExecutionServiceForStatus()(implicit executionContext: ExecutionContext): Future[ExecutionServiceStatusResponse] = {
+    datasource.inTransaction { dataAccess =>
+      dataAccess.workflowQuery.listWorkflowRecsForSubmissionAndStatuses(submissionId, WorkflowStatuses.runningStatuses: _*)
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:107]: this is a pretty small query ... but I think it's doing a table scan. Potentially we would benefit from an index on the STATUS column.

Will mysql use the foreign key index (WORKFLOW.SUBMISSION_ID->SUBMISSION.ID) when querying on submissionId, without the join to the submission table?

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+    val entities = updatedEntitiesAndWorkspace.collect { case Left((Some(entity), _)) => entity }
+    if (entities.isEmpty) DBIO.successful(0)
+    else dataAccess.entityQuery.save(SlickWorkspaceContext(workspace), entities)
+  }
+
+  def attachOutputs(workspace: Workspace, workflowsWithOutputs: Seq[(WorkflowRecord, ExecutionServiceOutputs)], entitiesById: Map[UUID, Entity], outputExprepressions: Map[String, String]): Seq[Either[(Option[Entity], Option[Workspace]), (WorkflowRecord, Seq[AttributeString])]] = {
+    workflowsWithOutputs.map { case (workflowRecord, outputsResponse) =>
+      val outputs = outputsResponse.outputs
+
+      val attributes = outputExprepressions.map { case (outputName, attributeName) =>
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: spelling

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 12 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:107]: explain plan shows it is using the FK index:

```
MySQL [rawls]> explain select `ID`, `EXTERNAL_ID`, `SUBMISSION_ID`, `STATUS`, `STATUS_LAST_CHANGED`, `ENTITY_ID` from `WORKFLOW` where `SUBMISSION_ID` = 0x043AE84815084E74A7C027ECC3308B6C and STATUS in ("Submitted", "Running", "Aborting");
+----+-------------+----------+------+---------------+-----------+---------+-------+------+------------------------------------+
| id | select_type | table    | type | possible_keys | key       | key_len | ref   | rows | Extra                              |
+----+-------------+----------+------+---------------+-----------+---------+-------+------+------------------------------------+
|  1 | SIMPLE      | WORKFLOW | ref  | FK_WF_SUB     | FK_WF_SUB | 16      | const |   37 | Using index condition; Using where |
+----+-------------+----------+------+---------------+-----------+---------+-------+------+------------------------------------+
```

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+      } else {
+        Right((workflowRecord, attributes.collect { case Failure(t) => AttributeString(t.getMessage) }.toSeq))
+      }
+    }
+  }
+
+  def updateEntityAndWorkspace(entity: Entity, workspace: Workspace, workflowOutputs: Map[String, Attribute]): (Option[Entity], Option[Workspace]) = {
+    //Partition outputs by whether their attributes are entity attributes (begin with "this.") or workspace ones (implicitly; begin with "workspace.")
+    //This assumption (that it's either "this." or "workspace.") will be guaranteed by checking of the method config when it's imported; see DSDEEPB-1603.
+    //Yes I know this is a var but it's more terse this way.
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: :-1:

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 10 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: Be consistent for all 3 about collecting here vs. collect in the function - either way is fine

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+  def saveErrors(errors: Seq[(WorkflowRecord, Seq[AttributeString])], dataAccess: DataAccess) = {
+    DBIO.sequence(errors.map { case (workflowRecord, errorMessages) =>
+      dataAccess.workflowQuery.filter(_.id === workflowRecord.id).update(workflowRecord.copy(status = WorkflowStatuses.Failed.toString)) andThen
+        dataAccess.workflowQuery.saveMessages(errorMessages, workflowRecord.id)
+    })
+  }
+
+  private def getUserInfo = {
+    val expiresInSeconds: Long = Option(credential.getExpiresInSeconds).map(_.toLong).getOrElse(0)
+    if (expiresInSeconds <= 5*60) {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:275]: Should this be configurable?  If so, make a JIRA to do it later rather than bothering for now.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 5 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: ok

File: src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala [hunk 16]
```diff
@@ -25,229 +32,294 @@ class SubmissionMonitorSpec(_system: ActorSystem) extends TestKit(_system) with
     super.afterAll
   }
 
-  //actorSelection isn't super reliable, so this function waits for workflow actors to spin up and returns a map from workflowId -> wfActor
-  private def waitForWorkflowActors(workspaceContext: SlickWorkspaceContext, submission: Submission, subMonActor: TestActorRef[SubmissionMonitor]) = {
-    val wfActors = MMap[String, ActorRef]()
+  private def await[T](f: Future[T]): T = Await.result(f, 5 minutes)
 
-    val workflowsWithIds = runAndWait(workflowQuery.getWithWorkflowIds(workspaceContext, submission.submissionId))
+  "SubmissionMonitory" should "queryExecutionServiceForStatus success" in withDefaultTestDatabase { dataSource: SlickDataSource =>
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala]: spelling

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 14 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: (fixed, thanks!)

**Comment by jmthibault79** [comment]
:+1: with [three](https://github.com/broadinstitute/rawls/pull/383/files#r59604677) [small](https://github.com/broadinstitute/rawls/pull/383#discussion_r59606552) [comments](https://github.com/broadinstitute/rawls/pull/383#discussion_r59608165)

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 12 above)
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:107]: Try doing the explain, but leaving out the `SUBMISSION_ID = 0x043AE84815084E74A7C027ECC3308B6C` clause - in other words, just filtering by status. Explain shows it's doing a full scan on that. I suspect with both clauses, it filters first to the submissionId, then does a scan over the remainder of records. I don't know how much another index would optimize this, but I know for sure the statuses are not indexed.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 8 above)
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:179]: the actor receive log message just states that the actor is checking to see if a submission can be finished (""handling execution service response ..."). I'd like to see - not super urgent - a message that the monitor actually did close a submission, and with which status.

File: src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala (hunk 16 above)
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala]: Scala does have StringLike ... JavaScript has truthy/falsey ... SubmissionMonitory sounds like fun

**Comment by davidangb** [comment]
Overall, I am 👍 .

I have one comment inline about possibly adding a db index on WORKFLOW.STATUS. I have another to improve debug logging.

Tests pass and this seems to do what it should do. I am still grokking the tests and trying to understand coverage and use cases, but if we need to push this through, then ok.

As a follow-on Jira story - definitely not part of this ticket - I'd like to explore the possibility of the submission monitor looking for orphan workflows and/or submissions that have been in a non-terminal state for too long ... where "too long" is some configurable time period.

**Comment by davidangb** [comment]
@dvoet back to you for the handful of fixes mentioned, many of which are optional. Test coverage _looks_ good but I am still not confident I know all the possible cases. Could you also add TODOs and/or Jira tickets for missing test cases?

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala (hunk 12 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:107]: I am sure it would do a full table scan without the submission id but what does that prove? The query that we run uses the submission id.

Anyway, there is probably benefit to having it so we can query for all running workflows, etc.

**Comment by dvoet** [comment]
@davidangb what jiras do you want me to add? To consider whether or not we need more coverage? As far as I am concerned I worked hard to to be thorough so I don't think there is anything missing.

**Comment by dvoet** [comment]
I added GAWB-534 and GAWB-535 for the index and unit tests respectively

**Comment by dvoet** [comment]
@davidangb back to you for final approval. Feel free to squash and merge it yourself if I am not available.

**Comment by davidangb** [comment]
Great, thanks for those two jira tickets. 👍 will merge.
